[SYCL][ESIMD] Add tests on simd vector move constructor

yuriykoch opened the pull request:
Context variations are moved out as a test cases, with any action run directly
over the simd move constructor result, with no further instance copy or move.

There is also an extra verification of the case when copy constructor is called
instead of the move constructor, by using a test_proxy functionality.

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Add tests on simd vector move constructor

Context variations are moved out as a test cases, with any action run directly
over the simd move constructor result, with no further instance copy or move.

yuriykoch committed:
[SYCL][ESIMD] Ensure the move constructor invocation

The copy constructor is a default fallback if there is no move
constructor available

yuriykoch committed:
[SYCL][ESIMD] Fix source type for simd move constructor

yuriykoch committed:
[SYCL][ESIMD] Use parallel_for for simd<T,1> case

yuriykoch committed:
[SYCL][ESIMD] Make flag usage in kernel more explicit

yuriykoch committed:
[SYCL][ESIMD] Disable tests on simd move constructor

Currently simd vector uses a copy constructor as a fallback

yuriykoch committed:
[SYCL][ESIMD] Update test_proxy API signature

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Use the latest logger interface

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch commented:
/verify with intel/llvm#5132

yuriykoch commented:
@vasilytric Would you mind to take a look?

vasilytric commented on SYCL/ESIMD/api/functional/ctors/ctor_move.cpp:
```diff
@@ -0,0 +1,232 @@
+        const auto ptr_out = result.data();
+        const auto ptr_flags = flags_storage.data();
+
+        cgh.parallel_for<KernelName>(
+            range, [=](sycl::id<1> id) SYCL_ESIMD_KERNEL {
+              const auto work_item_index = id[0];
+              const DataT *const in = ptr_in + work_item_index;
+              const auto out = ptr_out + work_item_index;
+              const auto was_moved_flag = ptr_flags + work_item_index;
+              *was_moved_flag = run_check(in, out);
+            });
+      });
+      queue.wait_and_throw();
```
please leave a comment that described what happened in this code

yuriykoch replied:
Right, please take a look onto 1e54ab1e38961f2b66bd4b6e38acbb254f3856ed

vasilytric commented on SYCL/ESIMD/api/functional/ctors/ctor_move.cpp:
```diff
@@ -0,0 +1,232 @@
+  // instance, so that we could check it later
+  // Returns the flag that should be true only if the move constructor was
+  // actually called, to differentiate with the copy constructor calls
+  static bool run_check(const DataT *const in, DataT *const out) {
```
could you use the same names for `input data` and `output data` as they uses in [copy constructor](https://github.com/intel/llvm-test-suite/blob/intel/SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp#L62)

yuriykoch replied:
From the point of `run_check` it's an input data it works on. There is no need to know what logic is on the upper level.

Moreover, for the case of `simd<T,1>` such input data represents a `single element from the reference data`, not the `reference data` itself, so using `ref_data` instead of `in` here sounds somewhat misleading.

Does it address your concerns?

vasilytric replied:
> Does it address your concerns?

yes, thanks for the clarification

vasilytric commented on SYCL/ESIMD/api/functional/ctors/ctor_move.cpp:
```diff
@@ -0,0 +1,232 @@
+  static std::string get_description() { return "variable declaration"; }
+
+  template <typename SimdT, typename ActionT>
+  static void run(SimdT &&source, const ActionT &action) {
+    static_assert(
+        type_traits::is_nonconst_rvalue_reference_v<decltype(source)>);
+
+    const auto instance(std::move(source));
+    action(instance);
+  }
+};
+
+// Uses the rvalue in expression C++ context to call simd move constructor
+struct rval_in_expr {
+  static std::string get_description() { return "rvalue in expression"; }
+
+  template <typename SimdT, typename ActionT>
+  static void run(SimdT &&source, const ActionT &action) {
+    static_assert(
+        type_traits::is_nonconst_rvalue_reference_v<decltype(source)>);
+
+    SimdT instance;
+    instance = SimdT(std::move(source));
+    action(instance);
+  }
+};
+
+// Uses the function argument C++ context to call simd move constructor
+class const_ref {
+public:
+  static std::string get_description() { return "const reference"; }
+
+  template <typename SimdT, typename ActionT>
+  static void run(SimdT &&source, const ActionT &action) {
+    static_assert(
+        type_traits::is_nonconst_rvalue_reference_v<decltype(source)>);
+
+    action(SimdT(std::move(source)));
+  }
+};
```
thinks about moving struct, that defined contexts, into the top of this file

yuriykoch replied:
Done by 51a796566cff3d78d86b8dc1929dd336308e73be

yuriykoch committed:
[SYCL][ESIMD] Clarify multiple work items usage

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Refactor the order of functions in file

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Remove unnecessary namespace using directive

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch commented:
@v-klochkov Would you mind to take a look after holidays? This PR uses intel/llvm#5132

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_move.cpp:
```diff
@@ -0,0 +1,233 @@
+// TODO Remove the level_zero restriction once the test is supported on other
+// platforms
+// UNSUPPORTED: cuda, hip
+// XRUN: %clangxx -fsycl %s -fsycl-device-code-split=per_kernel -o %t.out
```
@vladimirlaz - what is BKM for such tests. Should they be compiled/run and report fail, or skip the test compilation/run completely?
My concern regarding no-compilation and no-run approach used above is that it is unclear if the test is good, if it is compilable and does not have syntax errors.

yuriykoch replied:
Fair point. Please take a look into the 37e5e50466269c3e6e913393cb41b305915a1bbf as a suggestion.
The other option is to use `RUN: %clangxx -DSKIP_VECTOR_LEN_32 -fsycl ...`. By doing so we would be able:

- to disable some checks in CI
- to enable all checks "by default", making it possible to switch them off one-by-one in case of multiple feature flags during debugging session - with no changes to the test file itself

Still I'm not sure if the `RUN: %clangxx -DSKIP_VECTOR_LEN_32` solution would work for every setup or use-case, therefore the hard-coded macro was used.
Any feedback appreciated.

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_move.cpp:
```diff
@@ -0,0 +1,233 @@
+
+    // Action to run over the simd move constructor result
+    const auto action = [&](const simd<DataT, NumElems> &instance) {
+      was_moved = instance.get_test_proxy().was_move_destination();
```
It is odd, but I cannot find the definition of 'was_move_destination()'. Would you please give me a link to it?

yuriykoch replied:
Sure, sorry for any possible inconvenience.
Both the `test_proxy` class and the `test_proxy::was_move_destination()` are from the intel/llvm#5132.
Here is a link to the exact [line](https://github.com/intel/llvm/blob/f0465df6ee94da25704d82902649e74b04f85dd8/sycl/include/sycl/ext/intel/experimental/esimd/detail/test_proxy.hpp#L99).

v-klochkov replied:
Thank you.

yuriykoch committed:
[SYCL][ESIMD] Enable test run

XFAIL remains so if there is PR with move constructors added the
CI will fail effectively specifying the need to change the tests
The feature flag was hard-coded into the source to disable test
freeze for any use case, with or without parsing the RUN
directives

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch commented:
/verify with intel/llvm#5132

yuriykoch commented:
/verify

yuriykoch commented:
Looks like the simple "/verify" didn't work for me.
Anyway, the https://github.com/intel/llvm/pull/5132 is merged to the `intel:sycl` branch already.

yuriykoch commented:
@v-klochkov BTW, am I right to suppose that the `XFAIL: *` statement will result in test success for compilation failure?
So the test can still successfully run against any `intel/llvm` branch that doesn't have the intel/llvm#5132 patch from the `intel/sycl` merged.

v-klochkov commented:
> @v-klochkov BTW, am I right to suppose that the `XFAIL: *` statement will result in test success for compilation failure? So the test can still successfully run against any `intel/llvm` branch that doesn't have the [intel/llvm#5132](https://github.com/intel/llvm/pull/5132) patch from the `intel/sycl` merged.

I believe, that is correct.
